File: cpp/74_digit_factorial_chains.cpp
#include <iostream>
#include <set>

int get_chain(int n){
    int fact[10] = {1, 1, 2, 6, 24, 120, 720, 5040, 40320, 362880};
    
    std::set<int> seen;
    while (seen.find(n) == seen.end()){
        seen.insert(n);
        int total = 0;
        for (auto c : std::to_string(n)) total += fact[c - '0'];
        n = total;
    }
    return seen.size();
}

int digit_factorial_chains(){
    int count = 0;
    for (int i = 1; i < 1'000'000; i++){
        if (get_chain(i) == 60) count++;
    }
    return count;
}

int main(){
    std::cout << digit_factorial_chains() << "\n";
}

File: c#/_74_DigitFactorialChains.cs
using System;
using System.Collections.Generic;

public class _74_DigitFactorialChains{
	public static int Euler74(){
		int res = 0;
		int[] fact = {
			1, 1, 2, 6, 24, 120, 720, 5040, 40320, 362880
		};
		for (int i = 1; i < 1_000_000; i++){
			int n = i;
			HashSet<int> seen = new HashSet<int>();
			while (n < 1_000_000 && !seen.Contains(n)){
				seen.Add(n);
				int sum = 0;
				while (n > 0){
					sum += fact[n % 10];
					n /= 10;
				}
				n = sum;
			}
			if (seen.Count == 60) res++;
		}
		return res;
	}
	
	public static void Main(){
		Console.WriteLine(Euler74());
	}
}
